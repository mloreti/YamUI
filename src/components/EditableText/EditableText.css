/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';

.y-editableText--clickable {
  color: $textColor__metadata;
}

.y-editableText--clickable:active,
.y-editableText--clickable:focus,
.y-editableText--clickable:hover {
  color: $textColor__secondary;
}

.y-editableText--clickable:active > .y-editableText--clickableText,
.y-editableText--clickable:focus > .y-editableText--clickableText,
.y-editableText--clickable:hover > .y-editableText--clickableText {
  text-decoration: underline;
}
