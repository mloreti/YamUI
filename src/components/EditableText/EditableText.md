### Notes for use
EditableText will inherit the text size of the nearest Block component.

### Examples

With text and max length:

```js { "props": { "data-description": "with text and maxLength" } }
const { TextSize } = require('../Block');

<Block textSize={TextSize.SMALL}>
  <EditableText text="A spreadsheet of business requirements" maxLength={40} />
</Block>
```

With prompt text and placeholder text:

```js { "props": { "data-description": "with promptText and placeHolder" } }
const { TextSize } = require('../Block');

<Block textSize={TextSize.SMALL}>
  <EditableText promptText="Click to edit..." placeHolder="Attachment description" />
</Block>
```

With onBeginEditing, onEndEditing and onUpdate callbacks:

```js { "props": { "data-description": "with callbacks" } }
const { TextSize } = require('../Block');

<Block textSize={TextSize.SMALL}>
  <EditableText
    text="Document description"
    onUpdate={action('onUpdate')}
    onBeginEditing={action('onBeginEditing')}
    onEndEditing={action('onEndEditing')}
  />
</Block>
```
