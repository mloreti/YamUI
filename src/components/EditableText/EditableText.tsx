/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import Clickable from '../Clickable';
import EditIcon from '../Icon/icons/Edit';
import TextField, { ITextField } from '../TextField';
import { KeyCodes } from '../../util/enums';
import './EditableText.css';

export interface EditableTextProps extends BaseComponentProps {
  /**
   * Current text.
   */
  text?: string;

  /**
   * The text to display when text is empty.
   */
  promptText?: string;

  /**
   * The TextField placeHolder text.
   */
  placeHolder?: string;

  /**
   * An optional max length for the description field when editing.
   */
  maxLength?: number;

  /**
   * Returns the new text string when updated. Triggered by enter keypress or blur event to close the textfield.
   */
  onUpdate?: ((text: string) => void);

  /**
   * Triggered when the user enters edit mode.
   */
  onBeginEditing?: (() => void);

  /**
   * Triggered when the user exits edit mode.
   */
  onEndEditing?: (() => void);
}

export interface EditableTextState {
  isEditing: boolean;
  editedValue: string;
}

/**
 * Displays text which can be edited on click.
 */
export default class EditableText extends React.Component<EditableTextProps, EditableTextState> {
  private textFieldRef: ITextField;

  constructor(props: EditableTextProps) {
    super(props);
    this.state = {
      isEditing: false,
      editedValue: '',
    };
  }

  public render() {
    const { editedValue, isEditing } = this.state;
    const { text, promptText, placeHolder, maxLength } = this.props;

    if (isEditing) {
      return (
        <TextField
          underlined={true}
          onChange={this.updateInternalEditedDescription}
          value={editedValue}
          placeHolder={placeHolder}
          maxLength={maxLength}
          componentRef={this.setTextFieldRef}
          onBlur={this.commitEdit}
          onKeyDown={this.onKeyDown}
        />
      );
    }

    return (
      <span>
        <Clickable onClick={this.handleEditClick}>
          <EditIcon /> {text || promptText}
        </Clickable>
      </span>
    );
  }

  private onKeyDown = (event: any) => {
    switch (event.which) {
      case KeyCodes.escape:
        event.preventDefault();
        this.cancelEdit();
        break;
      case KeyCodes.enter:
        event.preventDefault();
        this.commitEdit();
        break;
      default:
    }
  };

  private setTextFieldRef = (component: ITextField) => {
    this.textFieldRef = component;
  };

  private updateInternalEditedDescription = (description: string) => {
    this.setState(() => ({ editedValue: description }));
  };

  private setTextFieldFocus() {
    window.setTimeout(() => {
      this.textFieldRef && this.textFieldRef.focus && this.textFieldRef.focus();
    }, 0);
  }

  private handleEditClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    this.setState(() => ({ isEditing: true, editedValue: this.props.text }));
    this.setTextFieldFocus();
    this.props.onBeginEditing && this.props.onBeginEditing();
  };

  private cancelEdit = () => {
    this.setState(() => ({ isEditing: false, editedValue: '' }));
    this.props.onEndEditing && this.props.onEndEditing();
  };

  private commitEdit = () => {
    const { editedValue } = this.state;
    this.setState(() => ({ isEditing: false, editedValue: '' }));
    this.props.onUpdate && this.props.onUpdate(editedValue);
    this.props.onEndEditing && this.props.onEndEditing();
  };
}
