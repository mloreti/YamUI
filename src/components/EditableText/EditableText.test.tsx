/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { mount, shallow, ReactWrapper, ShallowWrapper } from 'enzyme';
import EditableText, { EditableTextProps } from './index';

describe('<EditableText />', () => {
  let rawComponent: React.ReactElement<EditableTextProps>;
  let component: ShallowWrapper<EditableTextProps>;
  let mountedComponent: ReactWrapper<EditableTextProps>;
  let fakeEvent: any;

  beforeEach(() => {
    fakeEvent = {
      preventDefault: jest.fn(),
      stopPropagation: jest.fn(),
    };
  });

  describe('with promptText', () => {
    beforeEach(() => {
      component = shallow(<EditableText promptText="PROMPT TEXT" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<EditableText className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all props except promptText', () => {
    let updateCallback: jest.Mock;
    let beginCallback: jest.Mock;
    let endCallback: jest.Mock;

    beforeEach(() => {
      updateCallback = jest.fn();
      beginCallback = jest.fn();
      endCallback = jest.fn();

      rawComponent = (
        <EditableText
          text="TEXT"
          placeHolder="PLACEHOLDER"
          maxLength={120}
          onUpdate={updateCallback}
          onBeginEditing={beginCallback}
          onEndEditing={endCallback}
        />
      );
    });

    describe('shallow', () => {
      beforeEach(() => {
        component = shallow(rawComponent);
      });

      it('matches its snapshot', () => {
        expect(component).toMatchSnapshot();
      });
    });

    describe('mounted', () => {
      beforeEach(() => {
        mountedComponent = mount(rawComponent);
      });

      describe('when clicked', () => {
        beforeEach(() => {
          mountedComponent.find('button.y-editableText--clickable').simulate('click', fakeEvent);
        });

        it('enters edit mode', () => {
          expect(mountedComponent.find('.y-editableText--textfield input').length).toBe(1);
        });

        it('triggers the onBeginEditing callback', () => {
          expect(beginCallback).toHaveBeenCalledTimes(1);
        });

        describe('when text is updated', () => {
          // XXX: THIS IS BROKEN
          beforeEach(() => {
            const input = mountedComponent.find('.y-editableText--textfield input');
            input.simulate('focus');
            input.getDOMNode().nodeValue = 'abc';
            input.simulate('blur');
          });

          it('exits edit mode', () => {
            expect(mountedComponent.find('.y-editableText--textfield').length).toBe(0);
          });

          it('triggers the onEndEditing callback', () => {
            expect(endCallback).toHaveBeenCalledTimes(1);
          });

          it('triggers the update callback with the correct value', () => {
            expect(updateCallback).toHaveBeenCalledWith('abc');
          });
        });
      });
    });
  });
});
