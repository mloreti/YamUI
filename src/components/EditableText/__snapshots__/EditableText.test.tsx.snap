// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableText /> with additional className matches its snapshot 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
     
    <span
      className="y-editableText--clickableText"
    />
  </Clickable>
</span>
`;

exports[`<EditableText /> with all props except promptText shallow matches its snapshot 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
     
    <span
      className="y-editableText--clickableText"
    >
      TEXT
    </span>
  </Clickable>
</span>
`;

exports[`<EditableText /> with promptText matches its snapshot 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    ariaLabel="PROMPT TEXT"
    className="y-editableText--clickable"
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
     
    <span
      className="y-editableText--clickableText"
    >
      PROMPT TEXT
    </span>
  </Clickable>
</span>
`;
